kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java

```diff
@@ -201,13 +204,54 @@ public CloseableIterator<FilteredColumnarBatch> getAddFilesAsColumnarBatches(
    * use the P and/or M from the hint.
    */
   protected Tuple2<Protocol, Metadata> loadTableProtocolAndMetadata(
-      Engine engine, Optional<SnapshotHint> snapshotHint, long snapshotVersion) {
+      Engine engine,
+      LogSegment logSegment,
+      Optional<SnapshotHint> snapshotHint,
+      long snapshotVersion) {
 
-    // Exit early if the hint already has the info we need
+    // Exit early if the hint already has the info we need.
     if (snapshotHint.isPresent() && snapshotHint.get().getVersion() == snapshotVersion) {
       return new Tuple2<>(snapshotHint.get().getProtocol(), snapshotHint.get().getMetadata());
     }
 
+    // Snapshot hit is not use-able in this case for determine the lower bound.
+    if (snapshotHint.isPresent() && snapshotHint.get().getVersion() > snapshotVersion) {
+      snapshotHint = Optional.empty();
+    }
+
+    // Finds the inclusive lower bound for CRC search.
+    // If the snapshot hint or checkpoint older than required version is present, we can use them as
+    // the lower bound for the CRC search.
+    List<Long> eligibleCheckpointVersions =
+        logSegment.checkpoints.stream()
+            .map(cp -> checkpointVersion(cp.getPath()))
+            .filter(v -> v <= snapshotVersion)
+            .collect(Collectors.toList());
+    long crcSearchLowerBound =
+        max(
+            asList(
+                snapshotHint.map(SnapshotHint::getVersion).orElse(0L) + 1,
```

Comment:
nit - maybe add an inline comment why we add +1? might not be clear to a future reader